first-party/jib-quarkus-extension-gradle/src/main/java/com/google/cloud/tools/jib/gradle/extension/quarkus/JibQuarkusExtension.java

```diff
@@ -0,0 +1,178 @@
+import com.google.cloud.tools.jib.plugins.extension.ExtensionLogger.LogLevel;
+import com.google.cloud.tools.jib.plugins.extension.JibPluginExtensionException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Verify;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.jvm.tasks.Jar;
+
+public class JibQuarkusExtension implements JibGradlePluginExtension<Void> {
+
+  private AbsoluteUnixPath appRoot = AbsoluteUnixPath.get("/app");
+  private List<String> jvmFlags = Collections.emptyList();
+
+  @Override
+  public Optional<Class<Void>> getExtraConfigType() {
+    return Optional.empty();
+  }
+
+  @Override
+  public ContainerBuildPlan extendContainerBuildPlan(
+      ContainerBuildPlan buildPlan,
+      Map<String, String> properties,
+      Optional<Void> config,
+      GradleData gradleData,
+      ExtensionLogger logger)
+      throws JibPluginExtensionException {
+    try {
+      logger.log(LogLevel.LIFECYCLE, "Running Quarkus Jib extension");
+      readJibConfigurations(gradleData.getProject());
+
+      Project project = gradleData.getProject();
+      Path buildDir = project.getBuildDir().toPath();
+      Jar jarTask = (Jar) project.getTasks().findByName("jar");
+      String jarName = jarTask.getArchiveFile().get().getAsFile().getName();
+      Path jar = buildDir.resolve(jarName.replaceAll("\\.jar$", "-runner.jar"));
+
+      if (!Files.isRegularFile(jar)) {
+        throw new JibPluginExtensionException(
+            getClass(),
+            jar + " doesn't exist; did you run the Qaurkus Gradle plugin ('quarkusBuild' task)?");
+      }
+
+      ContainerBuildPlan.Builder planBuilder = buildPlan.toBuilder();
+      planBuilder.setLayers(Collections.emptyList());
+
+      // dependency layers
+      addDependencyLayers(project, planBuilder, buildDir.resolve("lib"));
+
+      // Quarkus runner JAR layer
+      AbsoluteUnixPath appRootJar = appRoot.resolve("app.jar");
+      FileEntriesLayer jarLayer =
+          FileEntriesLayer.builder().setName("quarkus jar").addEntry(jar, appRootJar).build();
+      planBuilder.addLayer(jarLayer);
+
+      // Preserve extra directories layers.
+      buildPlan
+          .getLayers()
+          .stream()
+          .filter(layer -> layer.getName().startsWith("extra files"))
+          .forEach(planBuilder::addLayer);
+
+      // set entrypoint
+      List<String> entrypoint = new ArrayList<>();
```

Comment:
are there any jvm args in the quarkus plugin?

Reply:
Good point. Looking at the default Quarkus `Dockerfile.jvm`, I think we can set up various env vars and flags.

```Dockerfile
ENV LANG='en_US.UTF-8' LANGUAGE='en_US:en'

# Install java and the run-java script
# Also set up permissions for user `1001`
RUN microdnf install curl ca-certificates ${JAVA_PACKAGE} \
    && microdnf update \
    && microdnf clean all \
    && mkdir /deployments \
    && chown 1001 /deployments \
    && chmod "g+rwX" /deployments \
    && chown 1001:root /deployments \
    && curl https://repo1.maven.org/maven2/io/fabric8/run-java-sh/${RUN_JAVA_VERSION}/run-java-sh-${RUN_JAVA_VERSION}-sh.sh -o /deployments/run-java.sh \
    && chown 1001 /deployments/run-java.sh \
    && chmod 540 /deployments/run-java.sh \
    && echo "securerandom.source=file:/dev/urandom" >> /etc/alternatives/jre/lib/security/java.security

# Configure the JAVA_OPTIONS, you can add -XshowSettings:vm to also display the heap size.
ENV JAVA_OPTIONS="-Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager"

COPY target/lib/* /deployments/lib/
COPY target/*-runner.jar /deployments/app.jar

EXPOSE 8080
USER 1001
```

Reply:
Thinking again, I think we can take care of with README documentation. They seem either not important or too specific.